### DISCUSSION TITLE
Duplicate dependancies causes an Ambiguous reference compilation error?

### DISCUSSION THREAD

**Discussion by jeremyosterhoudt** [ikvmnet/ikvm#442] [Q&A] [answered]
A college an myself are attempting to convert an old project that compiled Apache Calcite and a few other libraries using an older version of IKVM  to the latest version of IKVM.  I'm not sure if we have run into a bug or if we are missing a property setting in our project.  However, when we compile the project we end up with the following error:

```shell
error CS0433: The type 'Range' exists in both 'com.google.common, Version=31.1.0.0, Culture=neutral, PublicKeyToken=13235d27fcbfff58' and 'io.grpc, Version=1.9.0.0, Culture=neutral, PublicKeyToken=13235d27fcbfff58' [/dev/calcite-ikvm/CalciteIKVM/CalciteIKVM/CalciteIKVM.csproj]
```

Here the example project I'm using to duplicate the error we get on the larger project:

```
<Project Sdk="Microsoft.NET.Sdk">
    <PropertyGroup>
        <TargetFramework>net6.0</TargetFramework>
        <ImplicitUsings>enable</ImplicitUsings>
        <Nullable>enable</Nullable>
        <MavenAdditionalRepositories>$(MavenAdditionalRepositories);myrepo=https://repository.jboss.org/nexus/content/groups/public</MavenAdditionalRepositories>
    </PropertyGroup>
    
    <ItemGroup>
        <MavenReference Include="org.apache.calcite:calcite-core" Version="1.34.0" />
        <MavenReference Include="com.clickhouse:clickhouse-jdbc" Version="0.4.6" />
        <MavenReference Include="javax.jms:jms" Version="1.1" />
    </ItemGroup>

    <ItemGroup>
      <PackageReference Include="IKVM" Version="8.7.1" />
      <PackageReference Include="IKVM.Maven.Sdk" Version="1.6.1" />
    </ItemGroup>
</Project>
```

```
using Range = com.google.common.collect.Range;

namespace CalciteIKVM;

public class Class1
{
    public void Test()
    {
        var range = Range.singleton(10);
        var x = range.hasLowerBound() ? range.lowerEndpoint() : range.upperEndpoint();
    }
}
```

We are testing on an `Apple Silicon based Mac`.  We've not tried on other architectures.

**Comment by wasabii** [comment]
That's a CSharp error, not an IKVM error. Your references compiled fine. There are however types with the same name in both assemblies. Chances are the are types with the same name in both original JARs. Why? I do not know. Check, though.

**Reply by jeremyosterhoudt** [reply]
The references come from `com.clickhouse:clickhouse-jdbc` and from `org.apache.calcite:calcite-core`.  They both reference `Guava` which references `google.commons` jar.  We've seen this on a couple of other classes where two (or more) maven references have underline dependancies on the same components like Apache Commons.

If we remove either the `com.clickhouse:clickhouse-jdbc` or the `org.apache.calcite:calcite-core` the project will compile without issue.

In the past we used to specify a list of individual jars to include and then run them though `ikvmc`.  We were trying to use the maven reference as way of not having to manually pull dependancies.  Is it recommend for situations like these to use the `IkvmReference` property instead?

**Reply by wasabii** [reply]
The issue is not that two packages reference the same jars. That is a non-issue, and exactly what Maven is intended to handle. The problem is two JARs have the same classes. The recommendation is to understand WHY they do, address that issue, and make a decision on that basis.

Possibilities:

1) One of the libraries is old, and references and old library where the classes used to be, while those classes have since been moved to a different library in later versions.
2) One of the libraries intended to embed a private version of the same classes, and intends for those to be used instead of the common one. This is what IKVM.Maven.Sdk does: each resulting assembly has a reference to the resulting assembly the Maven dependencies declare it should have. But now it's up to you to reference the correct class you intend to use in CSharp.
3) The upstream authors made some sort of error and you should open an issue with them.

**Reply by jeremyosterhoudt** [reply]
In the case that two packages reference the same dependency with different versions, is that handled by default with the IKVM.Maven.Sdk or would the C# project need to reference the specific dependency and version as maven package?

**Reply by wasabii** [reply]
It isn't two packages referencing the same dependency with different versions. It's two packages referencing two different dependencies with the same classes. Hence the error:  The type 'Range' exists in both 'com.google.common, Version=31.1.0.0, Culture=neutral, PublicKeyToken=13235d27fcbfff58' and 'io.grpc, Version=1.9.0.0, Culture=neutral, PublicKeyToken=13235d27fcbfff58' [/dev/calcite-ikvm/CalciteIKVM/CalciteIKVM/CalciteIKVM.csproj]


The type, Range, is in both 'com.google.common', and also in 'io.grpc'.  Two different packages. One class.

**Reply by jeremyosterhoudt** [reply]
Yes, I understand that.  The part that is confusing is these two packages can be brought into a Java project using gradle or maven and the issue will not be present.  The project simple imports `com.google.common.collect.Range` and everything is fine.  In C# because of the namespace collision it seems as if the references need to be brought in individually (which is fine).

**Reply by wasabii** [reply, marked as answer]
Well, the issue can bepresent, in that you have duplicate overlapping class names on your classpath in Java, and if you attempted to compile code against both of these JAR files at the same time, javac.exe would throw an error regarding the duplicate class name. When you type new Range in Java, it wouldn't know which Range you were talking about.

So here's what I've found: https://github.com/ClickHouse/clickhouse-java/tree/main/third-party-libraries/io.grpc

There's a package `com.clickhouse:io.grpc` in Maven, dependended on by the clickhouse stuff. It appears to be a repackage of gRPC, but configured differently (with JPMS support). So yeah, it overlaps. Which one do you intend to use?

In Java, at RUNTIME, it'll pick whichever one appears FIRST on the classpath. As opposed to compile time, where it'll error. But you still have control over that, even with Maven, or you can just let Maven randomlly pick one, so it's still sort of up to you.

Once you pick which Range class you want to use, what you'll want to do is add an Alias to it. Which we will need some doctoring of your project file for.

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/namespaces#148-qualified-alias-member

**Reply by jeremyosterhoudt** [reply]
I was able to solve it using an assembly alias like the following in the csproj:

```
<Target Name="ChangeAliasesOfStrongNameAssemblies" BeforeTargets="FindReferenceAssembliesForReferences;ResolveReferences">
    <ItemGroup>
        <ReferencePath Condition="'%(FileName)' == 'com.google.common'">
            <Aliases>GoogleCommon</Aliases>
        </ReferencePath>
    </ItemGroup>
</Target> 
```

Thanks for the information.  It was helpfully in trying to figure out how to resolve the conflicted namespace issue.

**Reply by wasabii** [reply]
Np. That was my goal, to sort of walk you through to that answer.

What we are missing though is built in support on IKVM.Maven for setting Aliases.
